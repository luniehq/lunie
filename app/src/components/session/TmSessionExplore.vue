<template>
  <SessionFrame :icon="`language`">
    <TmFormStruct :submit="onSubmit">
      <h2 class="session-title bottom-indent">Explore with any address</h2>

      <div v-if="session.addresses.length > 0" class="session-list">
        <div
          v-for="account in filteredAddresses"
          :key="account.address"
          :title="account.address"
          @click="exploreWith(account.address)"
        >
          <div class="tm-li-session">
            <div class="tm-li-session-icon">
              <i class="material-icons notranslate circle">{{
                getAddressIcon(account.type)
              }}</i>
            </div>
            <div class="tm-li-session-text">
              <div class="tm-li-session-title">
                <span>{{ account.address | formatAddress(12) }}</span>
                <p class="tm-li-session-subtitle">
                  {{ getAddressTypeDescription(account.type) }}
                </p>
              </div>
            </div>
            <div class="tm-li-session-icon">
              <i class="material-icons notranslate">arrow_forward</i>
            </div>
          </div>
        </div>
      </div>

      <div class="session-main">
        <TmFormGroup field-id="sign-in-name" field-label="Your Address">
          <TmField
            v-model.trim="address"
            type="text"
            placeholder
            vue-focus="vue-focus"
          />
          <TmFormMsg
            v-if="$v.address.$error && !$v.address.required"
            name="Address"
            type="required"
          />
          <TmFormMsg
            v-else-if="$v.address.$error && !$v.address.addressValidate"
            type="custom"
            :msg="addressError"
          />
          <TmFormMsg v-else-if="error" :name="error" type="custom" />
        </TmFormGroup>
      </div>
      <TmFormGroup
        class="field-checkbox"
        field-id="sign-up-warning"
        field-label
      >
        <div class="field-checkbox-input">
          <label class="field-checkbox-label" for="select-testnet">
            <input id="select-testnet" v-model="testnet" type="checkbox" />
            This is a testnet address
          </label>
        </div>
      </TmFormGroup>
      <div class="session-footer">
        <TmBtn value="Explore" />
      </div>
    </TmFormStruct>
  </SessionFrame>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import { required } from "vuelidate/lib/validators"
import TmBtn from "common/TmBtn"
import SessionFrame from "common/SessionFrame"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import { formatAddress } from "src/filters"

export default {
  name: `session-explore`,
  components: {
    SessionFrame,
    TmBtn,
    TmField,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
  },
  filters: {
    formatAddress,
  },
  data: () => ({
    address: ``,
    error: ``,
    addressError: undefined,
    testnet: false,
  }),
  computed: {
    ...mapState([`session`]),
    ...mapGetters([`network`, `networks`]),
    filteredAddresses() {
      const selectedNetwork = this.networks.find(
        ({ id }) => id === this.network
      )
      // handling query not loaded yet or failed
      if (!selectedNetwork) return []

      return this.session.addresses
        .filter((address) =>
          address.address.startsWith(selectedNetwork.address_prefix)
        )
        .slice(-3)
    },
  },
  mounted() {
    this.address = localStorage.getItem(`prevAddress`)
    // Check testnet checkbox if current network is a testnet
    this.testnet = this.networks.find(
      (network) => network.id === this.network
    ).testnet
  },
  methods: {
    async onSubmit() {
      this.error = null
      this.$v.$touch()
      if (this.$v.$error) return

      let networkOfAddress
      try {
        networkOfAddress = await this.$store.dispatch("getNetworkByAccount", {
          account: {
            address: this.address,
          },
          testnet: this.testnet,
        })
      } catch (error) {
        this.error = error.message
        return
      }

      // needs to be done first because if not we are logging into the current network
      await this.selectNetworkByAddress(networkOfAddress)
      this.$store.dispatch(`signIn`, {
        sessionType: `explore`,
        address: this.address,
        networkId: this.network,
      })

      localStorage.setItem(`prevAddress`, this.address)
      this.$router.push({
        name: "portfolio",
        params: {
          networkId: networkOfAddress.slug,
        },
      })
    },
    async selectNetworkByAddress(network) {
      this.$store.dispatch(`setNetwork`, network)
    },
    getAddressIcon(addressType) {
      if (addressType === "explore") return `language`
      if (addressType === "ledger") return `vpn_key`
      if (addressType === "extension") return `laptop`
      if (addressType === "local") return `phone_iphone`
    },
    getAddressTypeDescription(addressType) {
      if (addressType === "explore") return `Explore Mode`
      if (addressType === "ledger") return `Ledger Nano S`
      if (addressType === "extension") return `Lunie Browser Extension`
      if (addressType === "local") return `Local`
    },
    async exploreWith(address) {
      this.address = address
      await this.onSubmit()
    },
    async addressValidate(address) {
      try {
        await this.$store.dispatch("getNetworkByAccount", {
          account: {
            address,
          },
          testnet: this.testnet,
        })
        return true
      } catch (error) {
        this.addressError = error.message
        return false
      }
    },
  },
  validations() {
    return {
      address: {
        required,
        addressValidate: this.addressValidate,
      },
    }
  },
}
</script>
<style scoped>
.tm-li-session {
  display: flex;
  padding: 1rem;
  margin: 0.5rem 0;
  background-color: var(--app-fg);
  border-radius: 0.25rem;
}

.tm-li-session:hover {
  cursor: pointer;
  background: var(--app-fg-hover);
}

.tm-li-session-icon {
  display: flex;
  align-items: center;
  justify-content: center;
}

.tm-li-session-icon i {
  font-size: 1.25rem;
}

.tm-li-session-title {
  color: var(--bright);
  font-size: var(--h4);
  font-weight: 400;
  margin-top: -0.4rem;
}

.tm-li-session-subtitle {
  display: block;
  width: 100%;
  font-size: var(--h6);
  font-weight: 400;
  color: var(--dim);
}

.tm-li-session-text {
  flex: 1;
  display: flex;
  justify-content: center;
  flex-flow: column nowrap;
  padding: 0 1rem;
}

.material-icons.circle {
  border: 2px solid var(--dim);
  color: var(--dim);
  border-radius: 50%;
  padding: 0.5rem;
}

.button {
  min-width: 90px;
}
</style>
