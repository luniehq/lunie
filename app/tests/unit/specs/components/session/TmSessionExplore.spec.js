import { shallowMount, createLocalVue } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import TmSessionExplore from "session/TmSessionExplore"

describe(`TmSessionExplore`, () => {
  const localVue = createLocalVue()
  localVue.use(Vuelidate)

  let wrapper, $store
  const networks = [
    {
      id: "cosmos-hub-testnet",
      address_prefix: "cosmos",
      testnet: true,
    },
    {
      id: "cosmos-hub-mainnet",
      address_prefix: "cosmos",
      testnet: false,
      slug: "cosmos-hub",
    },
    {
      id: "terra-testnet",
      address_prefix: "terra",
      testnet: true,
    },
    {
      id: "kusama",
      address_prefix: "",
      testnet: false,
      type: "polkadot",
    },
  ]
  const addresses = [
    {
      address: `cosmos1z8mzakma7vnaajysmtkwt4wgjqr2m84tzvyfkz`,
      type: `explore`,
    },
    {
      address: `cosmos1unc788q8md2jymsns24eyhua58palg5kc7cstv`,
      type: `ledger`,
    },
    {
      address: `cosmos1vxkye0mpdtjhzrc6va5lcnxnuaa7m64khj8klc`,
      type: `extension`,
    },
    {
      address: `cosmos1epsszxwps8ayeusfh8ru995atagc05sslwesuy`,
      type: `local`,
    },
  ]

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(() => true),
      getters: {
        network: "cosmos-hub-testnet",
        networks,
      },
      state: {
        session: {
          address: ``,
          addresses,
        },
      },
    }

    wrapper = shallowMount(TmSessionExplore, {
      localVue,
      mocks: {
        $router: {
          push: jest.fn(),
        },
        $store,
      },
    })
  })

  it(`shows a form to sign in with an address`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should close the modal on successful login`, async () => {
    wrapper.setData({
      address: `cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx`,
    })
    wrapper.vm.$emit = jest.fn()
    $store.dispatch = () =>
      Promise.resolve({
        slug: "cosmos-hub",
      })
    await wrapper.vm.onSubmit()
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith({
      name: "portfolio",
      params: { networkId: "cosmos-hub" },
    })
  })

  it(`should dispatch getNetworkByAccount on submit`, async () => {
    wrapper.setData({
      address: `cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx`,
    })
    await wrapper.vm.onSubmit()
    expect($store.dispatch).toHaveBeenCalledWith(`getNetworkByAccount`, {
      account: {
        address: `cosmos1thyn8gfapk2d0zsp6dysn99ynhcs2y759kwznx`,
      },
      testnet: true,
    })
  })

  it(`should show the last account used`, () => {
    localStorage.setItem(`prevAddress`, `cosmos1xxx`)
    wrapper = shallowMount(TmSessionExplore, {
      localVue,
      mocks: {
        $store,
      },
    })
    expect(wrapper.vm.address).toBe(`cosmos1xxx`)
  })

  it(`should explore with a previously used address`, async () => {
    let address = `cosmos1z8mzakma7vnaajysmtkwt4wgjqr2m84tzvyfkz`
    await wrapper.vm.exploreWith(address)
    expect($store.dispatch).toHaveBeenCalledWith(`getNetworkByAccount`, {
      account: {
        address: `cosmos1z8mzakma7vnaajysmtkwt4wgjqr2m84tzvyfkz`,
      },
      testnet: true,
    })
  })

  it(`should set an error message if the address isn't recognised by Lunie`, async () => {
    let address = `cosmos1z8mzakma7vnaajysmtkwt4wgjqr2m84tzvyfkz`
    $store.dispatch = () => Promise.reject(new Error("Not recognised address"))
    await wrapper.vm.addressValidate(address)
    expect(wrapper.vm.addressError).toBe("Not recognised address")
  })

  it(`should check testnet checkbox if current network is a testnet`, async () => {
    expect(wrapper.vm.testnet).toBe(true)
  })
})
