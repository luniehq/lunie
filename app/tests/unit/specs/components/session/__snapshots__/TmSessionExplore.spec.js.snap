// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionExplore shows a form to sign in with an address 1`] = `
<sessionframe-stub
  icon="language"
>
  <tmformstruct-stub
    submit="function () { [native code] }"
    width=""
  >
    <h2
      class="session-title bottom-indent"
    >
      Explore with any address
    </h2>
     
    <div
      class="session-list"
    >
      <div
        title="cosmos1unc788q8md2jymsns24eyhua58palg5kc7cstv"
      >
        <div
          class="tm-li-session"
        >
          <div
            class="tm-li-session-icon"
          >
            <i
              class="material-icons notranslate circle"
            >
              vpn_key
            </i>
          </div>
           
          <div
            class="tm-li-session-text"
          >
            <div
              class="tm-li-session-title"
            >
              <span>
                cosmos…palg5kc7cstv
              </span>
               
              <p
                class="tm-li-session-subtitle"
              >
                
                Ledger Nano S
              
              </p>
            </div>
          </div>
           
          <div
            class="tm-li-session-icon"
          >
            <i
              class="material-icons notranslate"
            >
              arrow_forward
            </i>
          </div>
        </div>
      </div>
      <div
        title="cosmos1vxkye0mpdtjhzrc6va5lcnxnuaa7m64khj8klc"
      >
        <div
          class="tm-li-session"
        >
          <div
            class="tm-li-session-icon"
          >
            <i
              class="material-icons notranslate circle"
            >
              laptop
            </i>
          </div>
           
          <div
            class="tm-li-session-text"
          >
            <div
              class="tm-li-session-title"
            >
              <span>
                cosmos…a7m64khj8klc
              </span>
               
              <p
                class="tm-li-session-subtitle"
              >
                
                Lunie Browser Extension
              
              </p>
            </div>
          </div>
           
          <div
            class="tm-li-session-icon"
          >
            <i
              class="material-icons notranslate"
            >
              arrow_forward
            </i>
          </div>
        </div>
      </div>
      <div
        title="cosmos1epsszxwps8ayeusfh8ru995atagc05sslwesuy"
      >
        <div
          class="tm-li-session"
        >
          <div
            class="tm-li-session-icon"
          >
            <i
              class="material-icons notranslate circle"
            >
              phone_iphone
            </i>
          </div>
           
          <div
            class="tm-li-session-text"
          >
            <div
              class="tm-li-session-title"
            >
              <span>
                cosmos…gc05sslwesuy
              </span>
               
              <p
                class="tm-li-session-subtitle"
              >
                
                Local
              
              </p>
            </div>
          </div>
           
          <div
            class="tm-li-session-icon"
          >
            <i
              class="material-icons notranslate"
            >
              arrow_forward
            </i>
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="session-main"
    >
      <tmformgroup-stub
        fieldid="sign-in-name"
        fieldlabel="Your Address"
      >
        <tmfield-stub
          placeholder=""
          type="text"
          vue-focus="vue-focus"
        />
         
        <!---->
      </tmformgroup-stub>
    </div>
     
    <tmformgroup-stub
      class="field-checkbox"
      fieldid="sign-up-warning"
      fieldlabel=""
    >
      <div
        class="field-checkbox-input"
      >
        <label
          class="field-checkbox-label"
          for="select-testnet"
        >
          <input
            id="select-testnet"
            type="checkbox"
          />
          
          This is a testnet address
        
        </label>
      </div>
    </tmformgroup-stub>
     
    <div
      class="session-footer"
    >
      <tmbtn-stub
        customclass=""
        value="Explore"
      />
    </div>
  </tmformstruct-stub>
</sessionframe-stub>
`;
